feat(testing): Add bootstrapAsPromised

nicojs opened the pull request:
## Informations

Fixes #481 

Type | Migration
---|---
Feature | ??

## Description
See #481 

## Usage example

Example to use your feature and to improve the documentation after merging your PR:
```
import {bootstrapAsPromised} from "@tsed/testing";
class TestServer extends ServerLoader {
}

beforeEach(async () => {
   await bootstrapAsPromised();
});
```

## Todos

- [ ] Tests
- [ ] Coverage
- [ ] Example
- [ ] Documentation

Romakita commented on packages/testing/src/TestContext.ts:
```diff
@@ -0,0 +1,11 @@
+
+export class TestContext {
+  static injector: InjectorService | null;
+  static reset() {
```
Nice :)
But where the reset method is used ?

Currently, no function are exposed to reset the injector and we need to update the documentation to explain this behavior.
Proposal:
```typescript
import {bootstrap, resetInjector} from "@tsed/testing";
import {Server} from "../../Server";

describe('MyTest', () => {
    before(bootstrap(Server));
    after(resetInjector);
})
```
What do you think ?

nicojs replied:
Yeah, indeed. That was what I was thinking. If bootstrap does something static, than it should also be reset after a test run. Good catch!

Romakita commented:
@milewski I included you because, I think you use this package to run your tools.  ;)

nicojs commented on packages/testing/src/bootstrap.ts:
```diff
@@ -18,12 +17,33 @@ export function bootstrap(server: any, ...args: any[]) {
+ * @param args
+ * @returns {(done:Function)=>undefined}
+ */
+export function bootstrapAsPromised(server: any, ...args: any[]): Promise<void> {
```
If we did this, i would suggest doing this:

```ts
export async function bootstrap(server: any, ...args: any[]) {
  $log.stop();
  process.env.NODE_ENV = process.env.NODE_ENV || "test";
   const instance = new server(...args);
   TestContext.injector = instance.injector;
   
  return instance.start();
  };
}
```

However, both cases would be a breaking change, no?

As a side note: Why doing this here:`$log.stop();  process.env.NODE_ENV = process.env.NODE_ENV || "test";`? These side effects are not what one might expect, right?

Romakita replied:
To help you, I added unit test to the current bootstrap implementation (on production branch).

The new unit test should be like this:
```typescript
import {bootstrap} from "@tsed/testing";
import {expect} from "../../tools";

class FakeServer {
  static current: FakeServer;
  startServers: Function;

  injector =  "Injector";

  async start() {
    FakeServer.current = this;

    return this.startServers();
  }
}

describe("bootstrap()", () => {
  const fn = bootstrap(FakeServer);

  before(fn);

  it("should return a before function", () => {
    expect(fn).to.be.a("function");
  });

  it('should attach injector instance to TestContext", () => {
      expect(TestContext.injector).to.eq('Injector');
  });

  it("should replace FakeServer.startServers by a stub()", () => {
    expect(FakeServer.current.startServers).to.be.a("Function");
    expect(FakeServer.current.startServers()).to.be.an.instanceOf(Promise);
  });
});
```

See you,
Romain :)

nicojs replied:
That's awesome! Thanks! I will definitely add/alter the unit tests. Still having troubles with running stuff locally. Will look into that.

> I'm not sure if my proposal broke the current code, because right now all test unit framework support promise

It will _definitely_ break existing code. For example here: https://github.com/stryker-mutator/stryker-dashboard/blob/master/packages/website-backend/test/helpers/TestServer.ts#L112. But that's fine. It is a breaking change, I'll leave it up to you to choose a correct version number and changelog.

I agree with your sentiment and change the code as you requested.

Romakita replied:
Ok it's a breaking change, you just have to write the right commit message and the deploy script will apply the right number version ;)

Example:
```
feat(test): My title

Optional body text
Closes: #487

BREAKING CHANGE: Reason of the breaking change
```
I'll create a new branch 5.x. It's a good time for me also to remove deprecated code ;)
So no problem to introduce breaking change :D 

See you,
Romain

nicojs replied:
Ok, I love conventional commit messages. We use the same for Stryker.

Are you doing squash commits? Or should I squash my commits myself?

Romakita replied:
I'll squash your commit if it's necessary ;)

nicojs commented:
I've force-pushed my changes. All review comments addressed. Let me know if you want me to change the history, at docs, etc.

Romakita commented:
Yes you can change the documentation ;)

I want to know what is the new usage of the @tsed/testing.
If i write a user story, it should be something like that:

> As a developer I want bootstrap my Server to test my code. Then I want to release the server or context initialized before the unit test.

I said that because you've created a TestContext, but do you want to expose this class for other developer or just for internal purpose ? 

Otherwise, Good job ;)

nicojs commented:
I've changed the documentation. Also updated the `inject` method to also work with promises. I've deleted the `Done` symbol, as it's no longer needed. If you want to do async stuff in your `inject` function, just return a promise and you're all set.

nicojs commented:
Build is green!! 🍏 📗 💚

Romakita commented:
Nice job @nicojs ;)

Romakita commented:
Ok. Excellente :D . I think, the next step (another PR) is to expose TestContext with a bootstrap method, inject service and reset context, to have only one API. 

```typescript
before(TestContext.boostrap(Server))
before(inject([MyService], (Service: MyService) =>{

});
after(TestContext.reset)
```
Other function like bootstrap and loadInjector, will just be deprecated in favor of TestContext API.

nicojs commented:
Sounds good.
